File: gateway/protocol/socket.py
import socket
from typing import Self, Any

# Files
FILE_MOVIES  = 0
FILE_CREDITS = 1
FILE_RATINGS = 2

# MSG kinds
MSG_BATCH = 0
MSG_FIN   = 1
MSG_ERR   = 2

def _recv_all(skt: socket.socket, n: int) -> bytearray:
    data = bytearray()
    while len(data) < n:
        read = skt.recv(n - len(data))
        if not read:
            raise ValueError("couldn't recv data through the socket")
        data.extend(read)
    return data

class Message:
    def __init__(self, kind: int, data: Any):
        self.kind = kind
        self.data = data

class CsvTransferStream:
    def __init__(self, skt: socket.socket):
        self._skt = skt

    def resource(self):
        file_id_bytes = _recv_all(self._skt, 1)
        file_id = int.from_bytes(file_id_bytes, "big")
        return ["movies", "credits", "ratings"][file_id]

    def recv(self) -> Message:
        msg_kind_bytes = _recv_all(self._skt, 1)
        msg_kind = int.from_bytes(msg_kind_bytes, "big")
        if msg_kind != MSG_BATCH:
            return Message(msg_kind, None)

        batch_size_bytes = _recv_all(self._skt, 4)
        batch_size = int.from_bytes(batch_size_bytes, "big")

        lines = []
        for _ in range(batch_size):
            len_bytes = _recv_all(self._skt, 4)
            len = int.from_bytes(len_bytes, "big")
            line = _recv_all(self._skt, len)
            lines.append(line)

        return Message(msg_kind, lines)

    def send(self, data: bytearray):
        self._skt.sendall(data)

    def close(self):
        self._skt.close()

class CsvTransferListener:
    def __init__(self, skt: socket.socket):
        self._skt = skt

    @classmethod
    def bind(cls, host: str, port: int, backlog: int = 0) -> Self:
        """
        Instanciates a new CsvTransferListener and binds it to the given address
        """
        self = cls(socket.socket(socket.AF_INET, socket.SOCK_STREAM))
        self._skt.bind((host, port))
        self._skt.listen(backlog)
        return self

    def accept(self) -> tuple[CsvTransferStream, "socket._RetAddress"]:
        """
        Blocks the calling thread until a new connection arrives
        """
        skt, addr = self._skt.accept()
        return CsvTransferStream(skt), addr

    def shutdown(self, how: int):
        self._skt.shutdown(how)

    def close(self):
        self._skt.close()

File: analyzer/gateway/protocol/socket.go
package protocol

import (
	"encoding/binary"
	"fmt"
	"io"
	"net"
)

const (
	MSG_BATCH = iota
	MSG_EOF
	MSG_ERR
)

// File identifiers
const (
	FILE_MOVIES  = 0
	FILE_CREDITS = 1
	FILE_RATINGS = 2
)

type Message struct {
	Kind int
	Data [][]byte
}

type CsvTransferStream struct {
	conn net.Conn
}

func NewCsvTransferStream(conn net.Conn) *CsvTransferStream {
	return &CsvTransferStream{conn: conn}
}

func (s *CsvTransferStream) Resource() (string, error) {
	fileIdBytes := make([]byte, 1)
	read, err := io.ReadFull(s.conn, fileIdBytes)
	if err != nil || read < len(fileIdBytes) {
		return "", fmt.Errorf("didn't read full %d bytes: %v", len(fileIdBytes), err)
	}
	fileId := int(fileIdBytes[0])

	files := []string{"movies", "credits", "ratings"}
	if fileId < 0 || fileId >= len(files) {
		return "", fmt.Errorf("invalid file ID received: %d", fileId)
	}

	return files[fileId], nil
}

func (s *CsvTransferStream) Recv() (*Message, error) {
	msgKindBytes := make([]byte, 1)
	read, err := io.ReadFull(s.conn, msgKindBytes)
	if err != nil || read < len(msgKindBytes) {
		return nil, fmt.Errorf("didn't read full %d bytes of msg kind: %v", len(msgKindBytes), err)
	}
	msgKind := int(msgKindBytes[0])

	if msgKind != MSG_BATCH {
		return &Message{Kind: msgKind, Data: nil}, nil
	}

	batchSizeBytes := make([]byte, 4)
	read, err = io.ReadFull(s.conn, batchSizeBytes)
	if err != nil || read < len(batchSizeBytes) {
		return nil, fmt.Errorf("didn't read full %d bytes of batch size: %v", len(batchSizeBytes), err)
	}
	batchSize := int(binary.BigEndian.Uint32(batchSizeBytes))

	lenBytes := make([]byte, 4)
	lines := make([][]byte, 0, batchSize)
	for range batchSize {
		read, err := io.ReadFull(s.conn, lenBytes)
		if err != nil || read < len(lenBytes) {
			return nil, fmt.Errorf("didn't read full %d bytes of length: %v", len(lenBytes), err)
		}
		lineLen := int(binary.BigEndian.Uint32(lenBytes))

		line := make([]byte, lineLen)
		read, err = io.ReadFull(s.conn, line)
		if err != nil || read < len(line) {
			return nil, fmt.Errorf("didn't read full %d bytes of line: %v", len(line), err)
		}
		lines = append(lines, line)
	}

	return &Message{Kind: msgKind, Data: lines}, nil
}

func writeAll(writer io.Writer, data []byte) error {
	written := 0
	for written < len(data) {
		n, err := writer.Write(data[written:])
		if err != nil {
			return err
		}
		written += n
	}
	return nil
}

func (s *CsvTransferStream) Send(data []byte) error {
	return writeAll(s.conn, data)
}

func (cts *CsvTransferStream) Close() error {
	return cts.conn.Close()
}

type CsvTransferListener struct {
	lis net.Listener
}

func Bind(host string, port int, backlog int) (*CsvTransferListener, error) {
	addr := net.JoinHostPort(host, fmt.Sprintf("%d", port))
	lis, err := net.Listen("tcp", addr)
	if err != nil {
		return nil, fmt.Errorf("couldn't bind through requested address %s: %v", addr, err)
	}
	return &CsvTransferListener{lis}, nil
}

func (s *CsvTransferListener) Accept() (*CsvTransferStream, net.Addr) {
	conn, err := s.lis.Accept()
	if err != nil {
		return nil, nil
	}
	return NewCsvTransferStream(conn), conn.RemoteAddr()
}

func (s *CsvTransferListener) Close() {
	s.lis.Close()
}
